src/objective/rank_objective.hpp

```diff
@@ -253,9 +277,41 @@ class LambdarankNDCG : public RankingObjective {
     }
   }
 
+  void UpdatePositionBiasFactors(const score_t* lambdas, const score_t* hessians) const override {
+    std::vector<double> bias_first_derivatives(num_position_ids_, 0.0);
+    std::vector<double> bias_second_derivatives(num_position_ids_, 0.0);
+    for (data_size_t i = 0; i < num_data_; i++) {
+      // accumilate first derivatives of utility w.r.t. position bias factors, for each position
+      bias_first_derivatives[positions_[i]] -= lambdas[i];
+      // accumilate second derivatives of utility w.r.t. position bias factors, for each position
+      bias_second_derivatives[positions_[i]] -= hessians[i];
+    }
+    #pragma omp parallel for schedule(guided)
+    for (data_size_t i = 0; i < num_position_ids_; i++) {
+      // do Newton-Rhapson step to update position bias factors
+      pos_biases_[i] += learning_rate_ * bias_first_derivatives[i] / (std::abs(bias_second_derivatives[i]) + 0.001);
+    }
+    LogDebugPositionBiasFactors();
+  }
+
   const char* GetName() const override { return "lambdarank"; }
 
  protected:
+  void LogDebugPositionBiasFactors() const {
+    std::stringstream message_stream;
+    message_stream << std::setw(15) << "position"
```

Comment:
Maybe use `Log::Debug` instead.

Reply:
Not sure I got this comment. We do use Log::Debug() eventually after creating and populating stringstream.

Reply:
Sorry, I did not notice that this is a `stringstream` and the message is already logged with `Log::Debug`.

Reply:
We can resolve this one.